bpo-34561: Switch to Munro & Wild "powersort" merge strategy.

tim-one opened the pull request:
<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
bpo-NNNN: Summary of the changes made
```

Where: bpo-NNNN refers to the issue number in the https://bugs.python.org.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `main`.

-->
First stab at this - work in progress, but so far so good. Still needs changes to listsort.txt, and, in listobject.c, at least reducing the max stack space needed (for remembering pending runs to be merged ).

<!-- issue-number: [bpo-34561](https://bugs.python.org/issue34561) -->
https://bugs.python.org/issue34561
<!-- /issue-number -->

tim-one committed:
bpo-34561: Switch to Munro & Wild "powersort" merge strategy.

tim-one committed:
Various changes. Most significantly, got rid of the new struct
member recording a run's starting index - that can be computed
cheaply on-the-fly, by storing instead the key array's base
address just once in the MergeState struct.

tim-one committed:
Close to done! Some code fiddling, changed comments to match the
code, and a new writeup for listsort.txt.

tim-one committed:
Typo repair.

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

tim-one commented on Misc/NEWS.d/next/Core and Builtins/2021-09-01-19-21-48.bpo-34561.uMAVA-.rst:
```diff
@@ -0,0 +1 @@
+List sorting now uses the merge-ordering strategy from Munro and Wild's ``powersort()``, Unlike the former strategy, this is provably near-optimal in the entropy of the distribution of run lengths. Most uses of ``list.sort()`` probably won't see a significant time difference, but may see significant improvements in cases where the former strategy was exceptionally poor. However, as these are all  ...
```
```suggestion
List sorting now uses the merge-ordering strategy from Munro and Wild's ``powersort()``. Unlike the former strategy, this is provably near-optimal in the entropy of the distribution of run lengths. Most uses of ``list.sort()`` probably won't see a significant time difference, but may see significant improvements in cases where the former strategy was exceptionally poor. However, as these are all fast linear-time approximations to a problem that's inherently at best quadratic-time to solve truly optimally, it's also possible to contrive cases where the former strategy did better.
```

tim-one committed:
Update Misc/NEWS.d/next/Core and Builtins/2021-09-01-19-21-48.bpo-34561.uMAVA-.rst

cfbolz commented on Objects/listobject.c:
```diff
@@ -1139,12 +1139,11 @@ sortslice_advance(sortslice *slice, Py_ssize_t n)
- *     32 * phi ** MAX_MERGE_PENDING
- * where phi ~= 1.618.  85 is ridiculouslylarge enough, good for an array
- * with 2**64 elements.
+ * For a list with n elements, thia needs at most floor(log2(n)) + 1 entries
```
typo: "this"? instead of "thia"

tim-one replied:
Thanks! Repaired. There are several Python implementations of the power computation in files attached to the bpo report. and the one here is a more-than-less mechanical translation of runstack.py's `powerloop()` function. I did check the C version on various cases, but agree "it would be nice" if that were an automated thing. But, it's not :wink:.

cfbolz commented on Objects/listobject.c:
```diff
@@ -1920,11 +1926,44 @@ merge_at(MergeState *ms, Py_ssize_t i)
+     * b = s1 + n1 + n2/2 = a + (n1 + n2)/2
+     */
+    Py_ssize_t a = 2 * s1 + n1;  /* 2*a */
+    Py_ssize_t b = a + n1 + n2;  /* 2*b */
```
bit of a nitpick, but I found this confusing when reading the code and trying to follow along with the comments: the math a and b from the comments having the same names as the code a and b, despite those storing 2*math a (and b respectively).

tim-one replied:
I added new comments to explain it. Does that help enough?

cfbolz replied:
They definitely help, but what tripped me up is really the reuse of the letter "a" to mean both "a" (in the comment) and "2*a" (in the code). so maybe the comment or the code could instead talk about x and y instead?

tim-one replied:
Thanks, but I'm disinclined to make more changes. The overwhelmingly important concept here is "midpoint", and I don't want to obscure it under layers of pedantic name proliferation. In my mind, the midpoints are named "a" and "b", period, and the shift in perspective by 1 bit position is a triviality much better overlooked than emphasized :wink:

cfbolz reviewed (commented):
not a super thorough review, but I did read through all the code and noticed a typo and a nitpick. (aside: it's a pity that we can't easily test those implementation-level C functions, it would be super nice to stick a random test for ``powerloop`` somewhere and compare it against a high level python implementation).

tim-one committed:
Comment changes to address good points raised in review.

tim-one committed:
Rewrote merge_collapse(), and renamed it to found_new_run(), to
reflect that it never merges the most recent run anymore. Putting
the new run on the stack before that function is called is a waste of
time now, and just complicated the function with now-silly code to pop
the run off at the start and push it back at the end. Now the caller
pushes it on the stack only after that function returns.

tim-one committed:
The starting-index argument to found_new_run wasn't actually used,
except in an assert. So get rid of it. Assert the run is adjacent
in its caller instead.
